protocol/smart-contract-versioning.md

```diff
@@ -0,0 +1,86 @@
+- `major` releases are to be used for changes that break the existing contract interface OR changes that modify the security model of a contract.
+
+Versioning for individual contracts:
+
+- A contract is only `X.Y.Z` on `develop` if it's been governance approved. If it's `X.Y.Z` before that, it must be on a branch. More on this below.
+- For contracts undergoing development, a `-beta.n` identifier must be appended to the version number.
+- For contracts in a release candidate state, an `-rc.n` identifier must be appended to the version number.
+- For contracts with feature-specific changes, a `+feature` identifier must be appended to the version number. See [ethereum-optimism/design-docs#8](https://github.com/ethereum-optimism/design-docs/pull/8) to learn more.
+- When making changes to a contract, always bump to the lowest possible version based on the specific change you are making. We do not want to e.g. optimistically bump to a major version, because protocol sequencing may change unexpectedly. Use these examples to know how to bump the version:
+  - Example 1: A contract is currently on `1.2.3`. We don't yet know when the next release of this contract will be. However, you are simply fixing typos in comments so you bump the version to `1.2.4-beta.1`.
+    - The next PR that makes in that same contract but makes no functional changes, so it bumps the version to `1.2.4-beta.2`.
+    - The last PR introduces a breaking change, which bumps the version from `1.2.4-beta.2` to `2.0.0-beta.1`.
+  - Example 2: A contract is currently on `2.4.7`. We know the next release of this contract will be a breaking change. Regardless, as you start development by fixing typos in comments, bump the version to `2.4.8-beta.1`. This is because we may end up putting out a release before the breaking change is added.
+    - Once you start working on the breaking change, bump the version to `3.0.0-beta.1`.
+    - Continue to bump the beta version as you make changes. When the contract is ready for release, bump the version to `3.0.0-rc.1`.
+- New contracts start at `0.y.z` and only become `1.0.0` when they are ready for production.
+
+Versioning for monorepo releases:
+
+- Monorepo releases continue to follow the `op-contracts/vX.Y.Z` naming convention.
+- The version used for the next release is determined by the highest version bump of any individual contract in the release.
+  - Example 1: The monorepo is at `op-contracts/v1.5.0`. Various tech debt is cleaned up in contracts, but no features are added, so all contracts only bump the patch version. The next monorepo release will be `op-contracts/v1.5.1`.
+  - Example 2: The monorepo is at `op-contracts/v1.5.1`. Legacy `ALL_CAPS()` getter methods are removed from a contract, causing that contract to bump the major version. The next monorepo release will be `op-contracts/v2.0.0`.
```

Comment:
> Intuitively, it doesn't make sense to me that removing a legacy method should cause the same version change as adding fault proofs.

Technically this is how it would work by following a strict semver, from https://semver.org/:
<img width="872" alt="image" src="https://github.com/ethereum-optimism/design-docs/assets/17163988/6956de7c-9be4-473e-89eb-2b0c6fd632ec">

So we just need to be thoughtful about breaking changes and bundle them together when appropriate (which we should be doing regardless of this proposal and how we handle semvers)

Reply:
Good response from semver.org, makes sense.